File: frontend/src/engine/adaptiveEngine.js
export class AdaptiveEngine {
  constructor(student, questions) {
    this.student = student;
    this.questions = questions;
  }

  getQuestionWeight(question) {
    const accuracy = this.student.getTopicAccuracy(question.topic);

    if (accuracy < 0.7) {
      return 3;
    }

    if (accuracy > 0.9) {
      return 1;
    }

    return 1;
  }

  createAdaptivePool() {
    const pool = [];

    for (const category of this.questions.category) {
      for (const question of category.questions) {
        const q = {
          ...question,
          topic: question.topic || category.name,
          difficulty: question.difficulty || 1,
          type: question.type || category.default_type || "choice",
        };

        const unlocked = this.student.getUnlockedDifficulty(q.topic);

        if (q.difficulty > unlocked) {
          continue;
        }

        q.weight = this.getQuestionWeight(q);

        pool.push(q);
      }
    }

    return pool;
  }

  selectQuestion(pool) {
    const totalWeight = pool.reduce((sum, q) => sum + q.weight, 0);

    let random = Math.random() * totalWeight;

    for (const question of pool) {
      random -= question.weight;

      if (random <= 0) {
        return question;
      }
    }

    return pool[pool.length - 1];
  }
}

File: backend/models/adaptive_engine.py
import random
from services.mastery_service import calculate_mastery


class AdaptiveEngine:
    def __init__(self, student, questions):
        self.student = student
        self.questions = questions

    def create_pool(self):
        pool = []

        for category in self.questions["category"]:
            for question in category["questions"]:
                q = {
                    **question,
                    "topic": question.get("topic", category["name"]),
                    "difficulty": question.get("difficulty", 1),
                }

                if "id" not in q:
                    q["id"] = category["name"] + "_" + str(len(pool))

                pool.append(q)

        return pool

    def select_question(self):
        pool = self.create_pool()

        mastery = calculate_mastery(
            self.student["history"]
        )

        weighted_pool = []

        for question in pool:
            topic = question["topic"]
            score = mastery.get(topic, 0)

            # Topic weighting
            if score < 50:
                weight = 5
            elif score < 80:
                weight = 3
            else:
                weight = 1

            # Difficulty weighting
            if question["difficulty"] == self.student["difficultyLevel"]:
                weight += 3

            weighted_pool.extend([question] * weight)

        return random.choice(weighted_pool)